### PR TITLE
[PeripheralBus] IOS GameController Support

### PR DESCRIPTION
**Pull request by fuzzard**
## Description
Implements a peripheralbus driver for darwin-embedded platforms.

As part of the peripheralbus implementation, this Implements an input driver for the Apple GCController API that supports both Extended and Micro button layouts.

Extended Gamecontroller supports ~~14~~ 16 buttons, 4 axis ~~(os13 potentially supports a further 2, will update when i have a controller to test with)~~
- Dpad (4 buttons)
- L/R Shoulder (2 buttons)
- L/R Trigger (currently buttons but provides axis data)
- A/B/X/Y
- Menu (>= ios13 required)
- Option (>= ios13 required)
- L/R Thumbsticks (2 axis each stick)
- L/R Thumbstick Press (>= ios12.1 required - ~~not implemented currently~~)

Micro Gamecontroller supports 6 buttons, 0 axis
- Dpad (4 buttons)
- A/X buttons
- Accelerometer/gyro data is possible, but not supported currently

Note: Micro support is entirely untested. The only micro controller im aware of is the
tvos Siri remote. TVOS input handling needs a complete overhaul for this to work on
that platform currently.

Todo:

- Multicontroller handling
- Convert Triggers to axis instead of buttons
- Investigate crash with Steelseries Nimbus Pause button
- Cleanup/refine Logging
- Investigate Micro Gyro/Accelerometer input

## Motivation and Context
Currently Game Controllers do nothing in iOS for Kodi. I have already sent some PR's (merged) for libretro based games. With the changes sent to peripheral.joystick and kodi-game-scripting repo's the beginning of libretro usage on ios aarch64 now can be tested and functional.

## How Has This Been Tested?
ios using an Xbox Controller and a Steelseries Nimbus

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [X] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/platform/darwin/ios-common/peripherals/PeripheralBusDarwinEmbedded.mm [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+        joystick->OnAxisMotion(event.DriverIndex(), event.AxisState());
+        break;
+      }
+      default:
+        break;
+    }
+  }
+  {
+    CSingleLock lock(m_critSectionStates);
+    // ToDo: Multiple controller handling
```
Comment (garbear) [inline xbmc/platform/darwin/ios-common/peripherals/PeripheralBusDarwinEmbedded.mm:191]:
```suggestion
    //! @todo Multiple controller handling
```

**Comment by garbear** [comment]
Changes to existing code look good. I'm not an objective-C jedi, but I was able to find the `for` loops for controllers, the `for` loops for buttons and axes, and the `if` statements to translate between APIs.

My only comment is that the dispatch code in the middle of [CPeripheralBusDarwinEmbedded::ProcessEvents()](https://github.com/xbmc/xbmc/pull/17675/files#diff-b40d04175464824066d45494f86423a6R155) looks like a duplicate of code in both android and add-on buses. Maybe a follow up commit that factors this out? Not a blocker for this PR.

**Comment by garbear** [comment]
And I *love* seeing line ratios like +1,267; −2.

600 lines added per single line changed is record territory!

**Comment by lrusak** [comment]
jenkins build this please

File: xbmc/platform/darwin/ios-common/peripherals/PeripheralBusDarwinEmbedded.mm (hunk 1 above)
Comment (fuzzard) [reply xbmc/platform/darwin/ios-common/peripherals/PeripheralBusDarwinEmbedded.mm]: Fixed all the Todo's up, cheers

**Comment by fuzzard** [comment]
I figure someones gotta put in LOC's to make up for all the stuff @lrusak is pulling out, haha

**Comment by fuzzard** [comment]
Linux test failure unrelated

**Pull request merged**